Nested mutable parameter forwarding breaks array indexing

## Description

When passing a mutable (`&`) array parameter to another function that also takes a mutable parameter, the inner function cannot use array indexing. The type becomes REFERENCE which doesn't support indexing.

## Reproduction

```ez
do outer(&arr [int]) {
    inner(arr)  // Forward the mutable param
}

do inner(&arr [int]) {
    arr[0] = 999  // ERROR: index operator not supported: REFERENCE
}

do main() {
    temp nums [int] = {1, 2, 3}
    outer(nums)
}
```

## Error

```
error[E3016]: index operator not supported: REFERENCE
```

## Expected Behavior

Nested mutable parameter forwarding should work correctly. The inner function should be able to index into the array just like the outer function can.

--- a/pkg/interpreter/evaluator_test.go
+++ b/pkg/interpreter/evaluator_test.go
@@ -1099,6 +1099,49 @@ str
 	testStringObject(t, evaluated, "modified")
 }
 
+func TestNestedMutableParameterForwarding(t *testing.T) {
+	// Test that mutable parameters work when forwarded through multiple functions
+	// This tests issue #338 - nested mutable parameter forwarding breaks array indexing
+	input := `
+do outer(&arr [int]) {
+	inner(arr)
+}
+
+do inner(&arr [int]) {
+	arr[0] = 999
+}
+
+temp nums [int] = {1, 2, 3}
+outer(nums)
+nums[0]
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 999)
+}
+
+func TestDeeplyNestedMutableParameterForwarding(t *testing.T) {
+	// Test mutable parameters forwarded through multiple levels
+	input := `
+do level1(&n int) {
+	level2(n)
+}
+
+do level2(&n int) {
+	level3(n)
+}
+
+do level3(&n int) {
+	n = 42
+}
+
+temp x int = 0
+level1(x)
+x
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 42)
+}
+
 // ============================================================================
 // String Interpolation Tests
 // ============================================================================

--- a/pkg/object/object.go
+++ b/pkg/object/object.go
@@ -125,8 +125,17 @@ func (r *Reference) Type() ObjectType { return REFERENCE_OBJ }
 func (r *Reference) Inspect() string  { return fmt.Sprintf("<ref %s>", r.Name) }
 
 // Deref returns the current value of the referenced variable
+// Recursively dereferences if the target is also a Reference (for nested mutable param forwarding)
 func (r *Reference) Deref() (Object, bool) {
-	return r.Env.Get(r.Name)
+	val, ok := r.Env.Get(r.Name)
+	if !ok {
+		return nil, false
+	}
+	// Chase through nested references
+	if ref, isRef := val.(*Reference); isRef {
+		return ref.Deref()
+	}
+	return val, ok
 }
 
 // SetValue updates the referenced variable's value
@@ -137,8 +146,13 @@ func (r *Reference) SetValue(val Object) bool {
 
 // updateRef updates a variable's value, traversing outer scopes if needed
 // This is used by Reference.SetValue to update through references
+// If the target is itself a Reference, chase through to the ultimate target
 func (e *Environment) updateRef(name string, val Object) bool {
-	if _, ok := e.store[name]; ok {
+	if current, ok := e.store[name]; ok {
+		// If current value is a Reference, chase through it
+		if ref, isRef := current.(*Reference); isRef {
+			return ref.SetValue(val)
+		}
 		e.store[name] = val
 		return true
 	}

